app/common/console_output.py

```diff
@@ -1,64 +1,81 @@
+import collections
+import zipfile
+
+from typing import BinaryIO, Optional
+
 from app.common.console_output_segment import ConsoleOutputSegment
+from app.util.exceptions import BadRequestError
+
 
+class ConsoleOutput:
+    """The console output of an atom"""
+
+    @classmethod
+    def from_plaintext(cls, path: str) -> 'ConsoleOutput':
+        """
+        :param path: Path to a plaintext file containing output
+        """
+        file = open(path, 'rb')
+        return cls(file=file)
+
+    @classmethod
+    def from_zipfile(cls, zip_path: str, path_in_archive: str) -> 'ConsoleOutput':
+        """
+        :param zip_path: Path to a zip archive containing an output file
+        :param path_in_archive: Path inside the zip archive to the file containing output
+        """
+        with zipfile.ZipFile(zip_path) as build_artifact:
+            file = build_artifact.open(path_in_archive)
+            return cls(file)
 
-class ConsoleOutput(object):
-    """
-    Represents the console output of an atom.
-    """
-    def __init__(self, path):
+    def __init__(self, file: BinaryIO):
         """
-        :param path: sys path to the console output file
-        :type path: str
+        This should normally only be called by static constructors.
+        :param file: open handle for the file containing console output
         """
-        self.path = path
+        self._file = file
 
-    def segment(self, max_lines=50, offset_line=None):
+    def segment(self, max_lines: int=50, offset_line: Optional[int]=None):
         """
-        Return a segment of the console output.
+        Return a segment of the console output. Note this closes the file handle so can only be called once
+        per ConsoleOutput object.
 
-        :type max_lines: int
+        :param max_lines: The maximum number of lines of output to return
         :param offset_line: The starting line number in the console output from which the segment should
             return content for. If set to None, then reads content starting from the end.
-        :type offset_line: int | None
-        :rtype: ConsoleOutputSegment
```

Comment:
Why remove this rtype docstring? Did you mean to add some other docstring in place of it?